onnx/defs/traditionalml/defs.cc

```diff
@@ -298,119 +298,129 @@ ONNX_ML_OPERATOR_SET_SCHEMA(
           }
 
-          // Check if the specified keys_* matches input type.
-          auto input_elem_type = ctx.getInputType(0)->tensor_type().elem_type();
-          if (keys_strings_result && input_elem_type != TensorProto::STRING) {
-            fail_shape_inference("Input type is not string tensor but key_strings is set");
-          }
-          if (keys_int64s_result && input_elem_type != TensorProto::INT64) {
-            fail_shape_inference("Input type is not int64 tensor but keys_int64s is set");
-          }
-          if (keys_floats_result && input_elem_type != TensorProto::FLOAT) {
-            fail_shape_inference("Input type is not float tensor but keys_floats is set");
+          int value_length, value_type;
+          std::tie(value_type, value_length) = getAttributeElementTypeAndLength(
+              ctx, {"values_as_tensor", "values_strings", "values_int64s", "values_floats"});
+          if (value_type == TensorProto::UNDEFINED) {
+            fail_shape_inference(
+                "At least one of values_as_tensor, values_strings, values_int64s, values_floats must be set.");
           }
 
-          // Load all values_* attributes.
-          std::vector<std::string> values_strings;
-          bool values_strings_result = getRepeatedAttribute(ctx, "values_strings", values_strings);
-          std::vector<int64_t> values_int64s;
-          bool values_int64s_result = getRepeatedAttribute(ctx, "values_int64s", values_int64s);
-          std::vector<float> values_floats;
-          bool values_floats_result = getRepeatedAttribute(ctx, "values_floats", values_floats);
+          int default_length, default_type;
+          std::tie(default_type, default_length) = getAttributeElementTypeAndLength(
+              ctx, {"default_as_tensor", "default_string", "default_int64", "default_float"});
+          if (default_type != TensorProto::UNDEFINED) {
+            if (value_type != default_type) {
+              fail_shape_inference(
+                  "The value type ",
+                  value_type,
+                  " and the default type ",
+                  default_type,
+                  " are different, which is not permitted for LabelEncoders.");
+            }
 
-          // Check if only one values_* attribute is set.
-          if (static_cast<int>(values_strings_result) + static_cast<int>(values_int64s_result) +
-                  static_cast<int>(values_floats_result) !=
-              1) {
-            fail_shape_inference("Only one of values_*'s can be set in label encoder.");
+            // Ensure default_as_tensor is a singleton if set
+            const AttributeProto* default_as_tensor = ctx.getAttribute("default_as_tensor");
+            if (default_as_tensor != nullptr && default_as_tensor->t().dims(0) != 1) {
+              fail_shape_inference("default_as_tensor must be a singleton if set.");
+            }
+          } else if (ctx.getAttribute("values_as_tensor") != nullptr) {
```

Comment:
Can we choose a rule similar to what already exists? 0 for all float types, -1 for all integer types, "_Unused" for string.

Reply:
Yes that is the plan if we infer a default here